compare-evidence-strings/compare.sh

```diff
@@ -1,130 +1,210 @@
+cut -f1 03.fields-and-strings.old | uniq -c | awk '$1>1 {print $2}' > 04.non-unique-fields.old \
+  & cut -f1 03.fields-and-strings.new | uniq -c | awk '$1>1 {print $2}' > 04.non-unique-fields.new \
+  & wait
+cat 04.non-unique-fields.old 04.non-unique-fields.new | sort -u > 05.all-non-unique-fields
 
-File 2
-${NEW_EVIDENCE_STRINGS}
-Total unique records: $(wc -l <03.fields-and-strings.new)
+echo "  Extract evidence strings with *non-unique* association fields into a separate group"
+join -t$'\t' -j 1 05.all-non-unique-fields 03.fields-and-strings.old > 06.non-unique.old \
+  & join -t$'\t' -j 1 05.all-non-unique-fields 03.fields-and-strings.new > 06.non-unique.new \
+  & wait
 
-The full diff between two files follows.
+echo "  Extract evidence strings with *unique* association fields into a separate group"
+# -v 2 means "print only records from file #2 which cannot be paired"
+# If records cannot be paired, it means their association fields are *not* in the list of duplicates
+join -t$'\t' -j 1 -v 2 05.all-non-unique-fields 03.fields-and-strings.old > 07.unique.old \
+  & join -t$'\t' -j 1 -v 2 05.all-non-unique-fields 03.fields-and-strings.new > 07.unique.new \
+  & wait
+
+echo "  Separate unique evidence strings into (deleted, common, new)"
+join -t$'\t' -j 1 07.unique.old 07.unique.new > 08.common \
+  & join -t$'\t' -j 1 -v 1 07.unique.old 07.unique.new > 08.deleted \
+  & join -t$'\t' -j 1 -v 2 07.unique.old 07.unique.new > 08.added \
+  & wait
 
-$(tail -n+5 "${DIFF_FILE}" | awk '{if ($0 !~ /@@/) {print $0 "\n"}}')
+
+
+########################################################################################################################
+echo "Compute differences for certain classes of evidence strings"
+
+echo "  Diff for evidence strings with *non-unique* association fields"
+compute_git_diff 06.non-unique.old 06.non-unique.new > 09.non-unique-diff
+
+echo "  Diff for evidence strings with *unique* association fields"
+cut -f2 08.common > 10.common.old & cut -f3 08.common > 10.common.new & wait
+compute_git_diff 10.common.old 10.common.new > 09.unique-diff
+
+echo "  Extract functional consequences for all changed evidence strings"
+extract_functional_consequences < 10.common.old > 11.consequences.old \
+  & extract_functional_consequences < 10.common.new > 11.consequences.new \
+  & wait
+
+echo "  Identify consequence type transitions"
+paste 11.consequences.old 11.consequences.new \
+  | awk -F$'\t' '$1 != $2' \
+  | sort > 12.consequences-transitions
+
+echo "  Compute frequencies of consequence type transitions"
+uniq -c 12.consequences-transitions | sort -k1,1rn > 99.consequences-transition-frequency
+
+
+
+########################################################################################################################
+echo "Produce the report"
+
+COLOR_RED='\033[0;31m'
+COLOR_GREEN='\033[0;32m'
+COLOR_RESET='\033[0m' # No Color
+export COLOR_RED COLOR_RESET
```

Comment:
Why not export COLOR_GREEN too?

Reply:
I simply missed that, thanks for noticing! Fixed in https://github.com/EBIvariation/eva-cttv-pipeline/pull/100/commits/82bbfda794e1405bb97e4cf79516bd9d4b888370.

Since the green was not exported, it seems like it works even without exporting. Which is a bit odd since my understanding was that `(...)` creates a new subshell which shouldn't have access to the environment of the main shell. But apparently it is not so. Anyway, I have kept exports for all of them since it seems safer